### ISSUE TITLE
How to get the 'old' value of an attribute?

### ISSUE THREAD

**Issue by MizterB** [custom-components/pyscript#433]
I would like to compare the new and old values of an attribute when it changes.  Specifically, I'd like to compare the current and previous brightness of a light when I detect the `brightness` attribute changing in a `@state_trigger`.

When I create a `@state_trigger` for an entity attribute, I do not have access to the 'old' value of that attribute inside the function.  I understand that the `domain.name.old.attr` form is available inside the decorator's expression, but that does not appear to carry into the function itself.  Instead, the `value` and `old_value` arguments always refer back to the entity state, not the entity attribute referenced in the trigger.

Is it possible to make this data available?

**Comment by craigbarratt** [comment]
You are correct that the `domain.name.old.attr` form only works in trigger expressions.

However, inside the function, `old_value.attr` should allow you to get the attributes of the prior entity state.

**Comment by MizterB** [comment]
Unfortunately, that's not the behavior I'm seeing with this code:
```python
@state_trigger("light.upstairs_hall_1_1_upstairs_hall_ceiling_motion_dimmer.brightness")
def jasco_motiondimmer_state(value=None, old_value=None):
    result = {}
    if value.brightness > 0:
        result["button_label"] = "up"
    elif value.brightness == 0:
        result["button_label"] = "down"
    log.debug(result)
```

When triggered, it throws
```
AttributeError: 'StateVal' object has no attribute 'brightness'
```

**Comment by craigbarratt** [comment]
Since your state trigger is the attribute `light.upstairs_hall_1_1_upstairs_hall_ceiling_motion_dimmer.brightness`, I believe `value` and `old_value` are the attribute value, not the entity (just based on quick inspection of the code; I haven't tested this).

Assuming my statement is correct, there are two different fixes.  Either change your trigger to:
```python
@state_trigger("light.upstairs_hall_1_1_upstairs_hall_ceiling_motion_dimmer")
```
or change your function to just use `value` and `old_value` instead of `value.brightness` etc.

**Comment by MizterB** [comment]
That's what I assumed, but it wasn't the behavior I was seeing - hence the issue.  I was triggering on the attribute, but the values reflected the entity 'state' value, not the 'attribute' one.

**Comment by MizterB** [comment]
Just retested to make sure...
```
@state_trigger("light.upstairs_hall_1_1_upstairs_hall_ceiling_motion_dimmer.brightness")
def jasco_motiondimmer_state(value=None, old_value=None):
    log.debug(f"value: {value}, old_value: {old_value}")
```
returns state, not attribute
```
value: on, old_value: off
```

**Comment by gibman** [comment]
@MizterB did you ever figure this one out ?
Here 3 years later I can confirm that on vers 1.7 the issue is still not solved.
I see the exact same problem.

In this example I try and get the new and old value for a media player domain entity.
value sadly just holds the entity's main state. eg on, off, idle.

At some point this must have worked as AI copilot / Claude as well as Gemini both claim that the value and old_value should contain the value of the attribute.

I can retrieve the new attribute value though. If I do this directly using state.getattr("my_attribute") . 
Ugly :/
But the old value is a no go.

`
@state_trigger("media_player.google.volume_level")
def on_vol_changed(value=None, old_value=None, **kw):
   #value just gives me the state of the entity and no the attribute itself - eg. volume_level
    volume_level_new = value #state.getattr("media_player.google").get("volume_level")
    volume_level_old =  ????
`

Background:
I spent a few days on trying to port my AppDaemon setup to PyScript.
But it seems I have to put this on the backburner for now.
AppDaemon though is a resource hog.
Using AI I was able to port around 12 appdaemon apps to pyscript.
But once I stumbled upon attribute listeners, problems started to happen.

**Comment by MizterB** [comment]
> [@MizterB](https://github.com/MizterB) did you ever figure this one out ? Here 3 years later I can confirm that on vers 1.7 the issue is still not solved. I see the exact same problem.

I don't believe I did.  Given the significant improvements to scripts/automation/templates since I opened this issue, I've moved nearly everything to HA native constructs.
